CRD webhook conversion

mbohlool opened the pull request:
This PR is ready for review. I plan to bring parts of it (specially the refactoring in first commits) into a new PR. What is intentionally missing from this PR to make it smaller is multi-scheme and multi-print-column etc. support that will be done in a separate PR.

The PR has an end to end test with a nice image that can be used as a reference to implement conversion webhooks (it has separate framework and user code). I will use that for documentation and kubebuilder support.


```release-note
Add support for CRD conversion webhook
```

--------------
Related PRs and Issues
--------------
- [Feature issue](https://github.com/kubernetes/features/issues/598)
- ~#67798: Refactor webhook calling code. Dependency of this PR. Already merged.~
- ~#67795: The API changes. Going through final API reviews. This PR is rebased on that.~
- ~#70211 : Multi version schema, additionalprintcolumn and subresources. Not a direct dependency of this PR but related. Current Status: ready to review~
- ~#70087: Support internal versions for CRDs (fix a bug in /scale). Already merged.~

mbohlool committed:
Example webhook implementation (used in E2E test)

mbohlool committed:
E2E Test

sttts commented on test/images/crd-conversion-webhook/scheme.go:
```diff
@@ -0,0 +1,37 @@
+}
+
+func addToScheme(scheme *runtime.Scheme) {
+	utilruntime.Must(corev1.AddToScheme(scheme))
```
where is this scheme used? Don't see any use.

mbohlool replied:
in the framework.go where: `codecs.UniversalDeserializer()` needs it as far as I remember.

mbohlool replied:
I see, you mean we do not need corev1 schemas to be added? We use schema to create the codec and codec to deserialize into a `ConversionReview` which has a `metav1.TypeMeta` that is defined in corev1.

sttts replied:
TypeMeta is no top-level type. It should work as it is without corev1.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/features/kube_features.go:
```diff
@@ -40,6 +40,13 @@ const (
+
+	// owner: @mbohlool
+	// alpha: v1.12
+	// beta: v1.13
```
we don't put future versions here.

mbohlool replied:
Done

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go:
```diff
@@ -0,0 +1,140 @@
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	"k8s.io/apiserver/pkg/admission/plugin/webhook/config"
```
this deserves a more generic location. But good that it is reused :+1:

mbohlool replied:
Sorry you looked at the old code. I just pushed a new one that move all of these to a new location, please take another look.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go:
```diff
@@ -136,6 +179,9 @@ const (
 	Terminating CustomResourceDefinitionConditionType = "Terminating"
+	// Valid means the CRD does pass validation. K8S objects are not validated on the GET path and it is
+	// possible to have invalid objects on downgrade/upgrade paths.
+	Valid CustomResourceDefinitionConditionType = "Valid"
```
odd condition.

Alternative would be something more specific: ValidConversion

mbohlool replied:
This should not be here. I removed it.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:
```diff
@@ -20,27 +20,66 @@ import (
+
+func NewCRDConverterFactory() (*CRDConverterFactory, error) {
+	converterFactory := &CRDConverterFactory{}
+	if utilfeature.DefaultFeatureGate.Enabled(apiextensionsfeatures.CustomResourceWebhookConversion) {
```
I would prefer to set this up even for disable feature gate, but not call the actual conversion depending on the gate. That's easier to reason about.

mbohlool replied:
I thought about it. The thing is if the creating of this factory fails for any reason, with turning feature gate off, we can bypass it but if we don't feature gate it, the conversions will be fail even if we don't use webhooks.

sttts replied:
When can it fail?

mbohlool replied:
right now `newWebhookConverterFactory` can fail if one of it's sub-calls fails. We can go into the details of those subcalls (`NewClientManager` and `NewDefaultAuthenticationInfoResolver`) but the fact that they can fail and even if we resolve failure for current code, it can change in future, make me believe that we should protect ourself from that by feature gate.

deads2k commented on pkg/kubeapiserver/admission/initializer.go:
```diff
@@ -18,8 +18,8 @@ package admission
 
 import (
 	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/util/webhook"
```
webhooks sound more client-y than machinery-y

deads2k commented:
Can you break `Refactor addmission webhook hook client to a util package` into a separate commit and move it into the client-go library instead of apimachinery?  I think I'm gathering that as I move down commits, you'll be re-using this package for the connection from teh apiextensions server.

deads2k commented:
The `types` commit appears to match the KEP.  Can you break it out into a separate pull to help focus the api-review ?

roycaihw commented on staging/src/k8s.io/apiextensions-apiserver/pkg/features/kube_features.go:
```diff
@@ -40,6 +40,12 @@ const (
+	// alpha: v1.12
+	//
+	// CustomResourceWebhookConversion defines the webhook conversion for Custom Resources.
+	CustomResourceWebhookConversion utilfeature.Feature = "CustomResourceWebhookConversion"
```
You also need to add `CustomResourceWebhookConversion` to `defaultKubernetesFeatureGates` in this file

mbohlool replied:
Yeah, it is unfortunate that we have the list of defaults duplicated for apiextensions. Done.

mbohlool commented:
/retest

sttts commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -176,7 +175,7 @@ func CreateServerChain(completedOptions completedServerRunOptions, stopCh <-chan
 		return nil, err
 	}
-	apiExtensionsServer, err := createAPIExtensionsServer(apiExtensionsConfig, genericapiserver.NewEmptyDelegate())
+	apiExtensionsServer, err := createAPIExtensionsServer(apiExtensionsConfig, serviceResolver, webhook.CreateWebhookAuthResolverWrapper(kubeAPIServerConfig.GenericConfig.LoopbackClientConfig, proxyTransport), genericapiserver.NewEmptyDelegate())
```
can we move `serviceResolver` and `authResolverWrapper` into the `apiExtensionsConfig`? Having them here and in the `New()` is a non-standard plumbing path reserved for very special cases.

sttts commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -535,35 +534,12 @@ func buildGenericConfig(
 		sharedInformers,
 		serviceResolver,
-		webhookAuthResolverWrapper,
+		webhook.CreateWebhookAuthResolverWrapper(genericConfig.LoopbackClientConfig, proxyTransport),
```
why is this not shared with the value above?

sttts commented on staging/src/k8s.io/api/admissionregistration/v1beta1/webhook_config_interface.go:
```diff
@@ -0,0 +1,60 @@
+
+// Implementing k8s.io/apiserver/pkg/util/webhook.ClientConfigInterface
+
+func (c *WebhookClientConfig) GetURL() *string {
```
Don't like to see this is an API package. Can we move that where it is actually used and turn it into a struct around `WebhookClientConfig`?

sttts replied:
See other comment: put it into a struct which embeds the api type. Then you can implement the interface everywhere.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go:
```diff
@@ -18,6 +18,16 @@ package apiextensions
+	"k8s.io/apimachinery/pkg/types"
+)
+
+type ConversionStrategyType string
```
go doc

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go:
```diff
@@ -51,6 +61,88 @@ type CustomResourceDefinitionSpec struct {
+	Conversion *CustomResourceConversion
+}
+
+type CustomResourceConversion struct {
```
go doc

sttts replied:
also everywhere below

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/defaults.go:
```diff
@@ -71,4 +71,9 @@ func SetDefaults_CustomResourceDefinitionSpec(obj *CustomResourceDefinitionSpec)
 	}
+	if obj.Conversion == nil {
+		obj.Conversion = &CustomResourceConversion{
+			Strategy: NopConverter,
```
what if Conversion is set, but Strategy is not. Can that happen?

mbohlool replied:
Strategy is not optional and if it has invalid values (e.g. empty string) it should be (and will be) a validation error.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go:
```diff
@@ -18,6 +18,16 @@ package v1beta1
+	"k8s.io/apimachinery/pkg/types"
+)
+
+type ConversionStrategyType string
```
godoc everywhere

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go:
```diff
@@ -56,6 +66,89 @@ type CustomResourceDefinitionSpec struct {
+// ServiceReference holds a reference to Service.legacy.k8s.io
+type ServiceReference struct {
+	// `namespace` is the namespace of the service.
+	// Required
```
do we write "Required" everywhere by convention? Never noticed this.

mbohlool replied:
Not sure about this and it does not mean anything to code generators.

mbohlool replied:
I prefer to keep these as this part is a copy from another already approved API type.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -110,13 +110,7 @@ func ValidateCustomResourceDefinitionSpec(spec *apiextensions.CustomResourceDefi
-	default:
-		allErrs = append(allErrs, field.NotSupported(fldPath.Child("scope"), spec.Scope, []string{string(apiextensions.ClusterScoped), string(apiextensions.NamespaceScoped)}))
-	}
+	allErrs = append(allErrs, validateEnumStrings(fldPath.Child("scope"), string(spec.Scope), []string{string(apiextensions.ClusterScoped), string(apiextensions.NamespaceScoped)}, true)...)
```
is that string cast really necessary?

mbohlool replied:
according to my go-compiler, yes

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -187,9 +181,31 @@ func ValidateCustomResourceDefinitionSpec(spec *apiextensions.CustomResourceDefi
 	}
 
+	if spec.Conversion != nil {
+		allErrs = append(allErrs, validateEnumStrings(fldPath.Child("conversion").Child("strategy"), string(spec.Conversion.Strategy), []string{string(apiextensions.NopConverter), string(apiextensions.WebhookConverter)}, true)...)
```
Worth a ValidateConversion?

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -187,9 +181,31 @@ func ValidateCustomResourceDefinitionSpec(spec *apiextensions.CustomResourceDefi
+		switch spec.Conversion.Strategy {
+		case apiextensions.WebhookConverter:
+			if spec.Conversion.Webhook == nil {
+				allErrs = append(allErrs, field.Required(fldPath.Child("spec", "conversion", "webhook"), "webhook field is required when strategy is set to webhook"))
```
no "spec" here.

sttts replied:
`set to Webhook`

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/webhook_config_interface.go:
```diff
@@ -0,0 +1,61 @@
+
+var _ webhook.ClientConfigInterface = &WebhookClientConfig{}
+
+func (c *WebhookClientConfig) GetURL() *string {
```
Do we need this interface based on versioned or internal types? 

As written above, please move this where the interface implementation is actually used. This is not API.

mbohlool replied:
We cannot define an interface for a type outside of its package, I find it cleaner to implement it here though I can have a wrapper type and implement interface for that. Which one you think is better?

sttts replied:
Just use embedding: `type fooInterfaceImplementation struct { webhookClientConfig apiextensionsv1beta1 }`. Works fine in Go.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go:
```diff
@@ -45,6 +45,7 @@ import (
 	_ "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
 	_ "k8s.io/apiextensions-apiserver/pkg/client/informers/externalversions"
 	_ "k8s.io/apiextensions-apiserver/pkg/client/informers/internalversion"
+	"k8s.io/apiserver/pkg/util/webhook"
```
import order

mbohlool replied:
Fixed and not the place to discuss it, but honestly I don't see any benefit with these import grouping when tools can auto import/order things in one list.

sttts replied:
These underscore imports are special. Init funcs....

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go:
```diff
@@ -119,7 +120,7 @@ func (cfg *Config) Complete() CompletedConfig {
 
 // New returns a new instance of CustomResourceDefinitions from the given config.
-func (c completedConfig) New(delegationTarget genericapiserver.DelegationTarget) (*CustomResourceDefinitions, error) {
+func (c completedConfig) New(serviceResolver webhook.ServiceResolver, authResolverWrapper webhook.AuthenticationInfoResolverWrapper, delegationTarget genericapiserver.DelegationTarget) (*CustomResourceDefinitions, error) {
```
move these into the config itself. I don't see a deeper reason to have have a non-standard plumbing path for these two.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:
```diff
@@ -20,27 +20,69 @@ import (
+	// if crd.Spec.Conversion is nil, that means the CRD created on an API server that do not have this field yet.
+	// That means we can default strategy to NopConverter.
+	Strategy := apiextensions.NopConverter
+	if crd.Spec.Conversion != nil {
```
How can this even happen? We do defaulting on read from etcd. Please remove this extra defaulting step.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:
```diff
@@ -20,27 +20,69 @@ import (
+// NewCRDConverterFactory creates a new CRDConverterFactory
+func NewCRDConverterFactory(serviceResolver webhook.ServiceResolver, authResolverWrapper webhook.AuthenticationInfoResolverWrapper) (*CRDConverterFactory, error) {
+	converterFactory := &CRDConverterFactory{}
+	if utilfeature.DefaultFeatureGate.Enabled(apiextensionsfeatures.CustomResourceWebhookConversion) {
```
we check for the gate below. Let's simplify this constructor by not checking the gate at all.

mbohlool replied:
`newWebhookConverterFactory` can fail and if we don't feature gate this, that means we can change construction behaviour without gating it. i.e. new code may fail initialization (if newWebhookConverterFactory fails) even though the gate is off.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go:
```diff
@@ -0,0 +1,179 @@
+
+	outGVK := unstructOut.GroupVersionKind()
+	if !c.validVersions[outGVK.GroupVersion()] {
+		return fmt.Errorf("request to convert CRD from an invalid group/version: %s", outGVK.String())
```
we convert CRs, not CRDs

mbohlool replied:
We had some old messages with this problem too, I fixed them all.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go:
```diff
@@ -248,3 +340,45 @@ type CustomResourceSubresourceScale struct {
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ConversionReview describes a conversion request/response.
+type ConversionReview struct {
```
We don't use internal ConversionRequests. Should we drop the types?

mbohlool replied:
It is the same for `AdmissionReview` as the internal version of that one is also not being used. I will drop it to see if anybody (runtime/compiler/reviewers) complains :)

sttts replied:
Keep a copy ;-)

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go:
```diff
@@ -0,0 +1,179 @@
+		return nil, fmt.Errorf("conversion request failed for %v: %v", in.GetObjectKind(), response.Response.Result.Message)
+	}
+
+	ret := response.Response.ConvertedObject.Object
```
do we have to check that `ConvertedObject` is not nil?

mbohlool replied:
ConvertedObject is not a pointer and required. The object inside can be nil though.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go:
```diff
@@ -0,0 +1,179 @@
+		return nil, fmt.Errorf("conversion webhook response was absent for %s", c.name)
+	}
+
+	if response.Response.Result != nil {
```
See comment in types.go

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -93,6 +94,8 @@ type crdHandler struct {
 	// CRD establishing process for HA clusters.
 	masterCount int
+
+	crdConverterFactory *conversion.CRDConverterFactory
```
strip crd prefix.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -427,7 +437,10 @@ func (r *crdHandler) getOrCreateServingInfoFor(crd *apiextensions.CustomResource
 
 	for _, v := range crd.Spec.Versions {
-		safeConverter, unsafeConverter := conversion.NewCRDConverter(crd)
+		safeConverter, unsafeConverter, err := r.crdConverterFactory.NewConverter(crd)
```
isn't one converter enough for all versions?

mbohlool replied:
This is a discussion from the previous PR (versioning PR), @liggitt wanted the converter object for each version, am I right @liggitt?

liggitt replied:
iirc, that was the discussion at https://github.com/kubernetes/kubernetes/pull/63830#discussion_r188359702

I didn't feel strongly about having a convertor per-version. The things I thought were important were:
* construction of the convertor happened when setting up servingInfo (once per CRD change, not once per custom resource handling request)
* the convertor had an explicit list of source/target versions it would accept conversion requests for (not just a list of target versions it would stomp to)

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler_test.go:
```diff
@@ -86,7 +86,11 @@ func TestConvertFieldLabel(t *testing.T) {
 			}
-			_, c := conversion.NewCRDConverter(&crd)
+			_, c, err := conversion.NewCRDConverter(&crd)
+
```
no empty line

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/options/options.go:
```diff
@@ -99,6 +103,22 @@ func (o CustomResourceDefinitionsServerOptions) Config() (*apiserver.Config, err
 }
 
+// New creates a new generic API service object using this option.
+func (o CustomResourceDefinitionsServerOptions) New() (*genericapiserver.GenericAPIServer, error) {
```
Don't put this here in the options. The logic below belong into `completeConfig.New` https://github.com/kubernetes/kubernetes/blob/2bf66c377d60276e017e78ab9dbc0490f26baa17/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go#L122. Why do we externalize it?

mbohlool replied:
Is it look good now that we moved those fields into config?

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/server.go:
```diff
@@ -51,14 +49,9 @@ func NewServerCommand(out, errOut io.Writer, stopCh <-chan struct{}) *cobra.Comm
-	}
-
-	server, err := config.Complete().New(genericapiserver.NewEmptyDelegate())
+	server, err := o.New()
```
leave the config+options logic here.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/initializer/initializer.go:
```diff
@@ -17,16 +17,15 @@ limitations under the License.
 	"k8s.io/apiserver/pkg/admission"
-	webhookconfig "k8s.io/apiserver/pkg/admission/plugin/webhook/config"
+	"k8s.io/apiserver/pkg/util/webhook"
+	"net/url"
```
import order

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go:
```diff
@@ -25,21 +25,22 @@ import (
 
 	admissionv1beta1 "k8s.io/api/admission/v1beta1"
-	"k8s.io/api/admissionregistration/v1beta1"
+	v1beta1 "k8s.io/api/admissionregistration/v1beta1"
```
no need for this alias

sttts commented on staging/src/k8s.io/apiserver/pkg/util/webhook/client.go:
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2017 The Kubernetes Authors.
+Copyright 2018 The Kubernetes Authors.
```
2017

sttts commented on staging/src/k8s.io/apiserver/pkg/util/webhook/client.go:
```diff
@@ -14,32 +14,40 @@ See the License for the specific language governing permissions and
+
 var (
-	ErrNeedServiceOrURL = errors.New("webhook configuration must have either service or URL")
+	errNeedServiceOrURL = errors.New("webhook configuration must have either service or URL")
```
Why is this a var? I only see one reference.

mbohlool replied:
This is moved code. fixed.

sttts commented on staging/src/k8s.io/apiserver/pkg/util/webhook/client.go:
```diff
@@ -14,32 +14,40 @@ See the License for the specific language governing permissions and
+type ClientConfigInterface interface {
+	GetURL() *string
+	GetCABundle() []byte
+	GetServiceName() *string
```
why all these string pointers? odd.

mbohlool replied:
Because Service is optional.

sttts replied:
We usually use empty strings for unset strings.

mbohlool replied:
Can any of those be a valid empty string? like a valid Path or Service Namespace?

mbohlool replied:
Actually we do not use empty string everywhere for that (and I think API convention is to user pointer for optional strings we just don't enforced that well). Example of that is Service.Path in this same struct that is a string pointer.

sttts commented on test/e2e/apimachinery/crd_conversion_webhook.go:
```diff
@@ -0,0 +1,295 @@
+
+		// Make sure the relevant provider supports admission webhook
+		framework.SkipUnlessServerVersionGTE(serverCRDConversionWebhookVersion, f.ClientSet.Discovery())
+		framework.SkipUnlessProviderIs("gce", "gke", "local")
```
why?

mbohlool replied:
I also have this question. Got this from admission webhooks. @caesarxuchao do you know why?

caesarxuchao replied:
aws doesn't support webhook back then, not sure if it's necessary anymore.

sttts commented on test/e2e/apimachinery/crd_conversion_webhook.go:
```diff
@@ -0,0 +1,295 @@
+	BeforeEach(func() {
+		client = f.ClientSet
+
+		// Make sure the relevant provider supports admission webhook
```
s/admission/conversion/

sttts commented on test/e2e/apimachinery/crd_conversion_webhook.go:
```diff
@@ -0,0 +1,295 @@
+})
+
+func cleanCRDWebhookTest(client clientset.Interface, namespaceName string) {
+	_ = client.CoreV1().Services(namespaceName).Delete(serviceCRDName, nil)
```
isn't the namespace deleted anyway?

mbohlool replied:
I guess so though we do this for admission webhooks here: https://github.com/kubernetes/kubernetes/blob/733a381ec42ede15cd0499bc502a0cba6612f3ca/test/e2e/apimachinery/webhook.go#L1067

Does it worth investigating? I mean no harm removing them manually, right?

sttts commented on test/images/crd-conversion-webhook/Dockerfile:
```diff
@@ -0,0 +1,18 @@
+# Copyright 2017 The Kubernetes Authors.
```
2018

sttts commented on test/images/crd-conversion-webhook/config.go:
```diff
@@ -0,0 +1,51 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
2018

sttts commented on test/images/crd-conversion-webhook/converter/converter_test.go:
```diff
@@ -0,0 +1,17 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
2018

sttts commented on test/images/crd-conversion-webhook/converter/example_converter.go:
```diff
@@ -0,0 +1,77 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
2018

sttts commented on test/images/crd-conversion-webhook/converter/scheme.go:
```diff
@@ -0,0 +1,37 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
2018

sttts commented on test/images/crd-conversion-webhook/converter/scheme.go:
```diff
@@ -0,0 +1,37 @@
+}
+
+func addToScheme(scheme *runtime.Scheme) {
+	utilruntime.Must(corev1.AddToScheme(scheme))
```
why core?

mbohlool replied:
Didn't I already answered this? I remember something. ConversionReview has metav1.TypeMeta in it which is part of corev1, right?

mbohlool replied:
The Schema is used to deserialize conversionReview.

sttts replied:
I don't see a reason for corev1 here. TypeMeta does not need it. Maybe in admission a pod was sent? Here it isn't.

sttts commented on test/images/crd-conversion-webhook/main.go:
```diff
@@ -0,0 +1,52 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
2018

mbohlool commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go:
```diff
@@ -263,3 +356,47 @@ type CustomResourceSubresourceScale struct {
+	// ConvertedObject is the converted version of request.Object if the Result is nil.
+	// +optional
+	ConvertedObject runtime.RawExtension `json:"convertedObject" protobuf:"bytes,2,name=convertedObject"`
+	// Result contains extra details into why a conversion request was failed. If it is not nil, it means
```
I agree with you, we do have a `Result` in `AdmissionResponse` with the same meaning. I will call this ErrorResult.

mbohlool commented:
@deads2k Sure, I was waiting for a round of review to create new sub-PRs. Also a good idea to create a type change PR. Here are both PRs: #67798, #67795

@sttts I think I addressed your comments. PTAL.

sttts commented on cmd/kube-apiserver/app/apiextensions.go:
```diff
@@ -71,12 +74,14 @@ func createAPIExtensionsConfig(
 }
 
-func createAPIExtensionsServer(apiextensionsConfig *apiextensionsapiserver.Config, delegateAPIServer genericapiserver.DelegationTarget) (*apiextensionsapiserver.CustomResourceDefinitions, error) {
+func createAPIExtensionsServer(apiextensionsConfig *apiextensionsapiserver.Config, serviceResolver webhook.ServiceResolver, authResolverWrapper webhook.AuthenticationInfoResolverWrapper, delegateAPIServer genericapiserver.DelegationTarget) (*apiextensionsapiserver.CustomResourceDefinitions, error) {
```
the two args are not used anymore.

sttts replied:
This still holds, doesn't it?

sttts commented on cmd/kube-apiserver/app/apiextensions.go:
```diff
@@ -71,12 +74,14 @@ func createAPIExtensionsConfig(
 			CRDRESTOptionsGetter: apiextensionsoptions.NewCRDRESTOptionsGetter(etcdOptions),
 			MasterCount:          masterCount,
+			AuthResolverWrapper:  authResolverWrapper,
+			ServiceResolver:      serviceResolver,
```
:+1:

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -187,6 +181,35 @@ func ValidateCustomResourceDefinitionSpec(spec *apiextensions.CustomResourceDefi
+// ValidateCustomResourceDefinitionSpecUpdate statically validates
+func ValidateCustomResourceConversion(conversion *apiextensions.CustomResourceConversion, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if conversion != nil {
```
return early on nil

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -187,6 +181,35 @@ func ValidateCustomResourceDefinitionSpec(spec *apiextensions.CustomResourceDefi
+		switch conversion.Strategy {
+		case apiextensions.WebhookConverter:
+			if conversion.Webhook == nil {
+				allErrs = append(allErrs, field.Required(fldPath.Child("webhook"), "webhook field is required when strategy is set to Webhook"))
```
do you have to repeat "webhook" in the error text?

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go:
```diff
@@ -78,6 +79,9 @@ type ExtraConfig struct {
 	// the CRD Establishing will be hold by 5 seconds.
 	MasterCount int
+
+	ServiceResolver     webhook.ServiceResolver
```
go doc

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:
```diff
@@ -20,27 +20,62 @@ import (
-// NewCRDConverter returns a new CRD converter based on the conversion settings in crd object.
-func NewCRDConverter(crd *apiextensions.CustomResourceDefinition) (safe, unsafe runtime.ObjectConvertor) {
+// CRDConverterFactory is the factory for all CRD converters.
+type CRDConverterFactory struct {
```
nit: CRConverter actually.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:
```diff
@@ -20,27 +20,62 @@ import (
 	}
-	return &safeConverterWrapper{unsafe}, unsafe
+
+	return nil, nil, fmt.Errorf("unknown conversion strategy \"%s\" for CRD %s", crd.Spec.Conversion.Strategy, crd.Spec.Group)
```
`%q` for `\"%s\"`

sttts replied:
and better use the crd name. The group alone does not say much.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go:
```diff
@@ -0,0 +1,180 @@
+	return &webhookConverterFactory{clientManager}, nil
+}
+
+// nopConverter is a converter that only sets the apiVersion fields, but does not real conversion.
```
webhookConverter

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go:
```diff
@@ -0,0 +1,180 @@
+	toGVK, ok := target.KindForGroupVersionKinds([]schema.GroupVersionKind{fromGVK})
+	if !ok {
+		// TODO: should this be a typed error?
+		return nil, fmt.Errorf("%v is unstructured and is not suitable for converting to %q", fromGVK.String(), target)
```
is this an error that should ever happen?

sttts commented on staging/src/k8s.io/client-go/util/webhook/client.go:
```diff
@@ -167,13 +171,13 @@ func (cm *ClientManager) HookClient(h *v1beta1.Webhook) (*rest.RESTClient, error
 	}
 
-	u, err := url.Parse(*h.ClientConfig.URL)
+	u, err := url.Parse(*h.GetURL())
```
ugly to call the getter multiple times, especially for pointers.

dims commented:
@sttts @mbohlool this is for 1.12 right? can you please mark the milestone

erictune reviewed (approved):
I approve of the schema as an api-reviewer.  I am not qualified to review the rest of the code, but approve of the overall direction.

spiffxp commented:
ping, this needs a rebase

shubheksha commented:
ping @mbohlool, could you please rebase? Thanks!

mbohlool committed:
Update generated files

mbohlool commented on test/images/crd-conversion-webhook/converter/example_converter.go:
```diff
@@ -0,0 +1,95 @@
+	convertedObject := Object.DeepCopy()
+	fromVersion := Object.GetAPIVersion()
+
+	// API Server may send a conversion request for an empty object to test the conversion stack before starting
```
@liggitt This is a stack trace in case you are interested to see why:

```
k8s.io/kubernetes/vendor/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion.(*webhookConverter).ConvertToVersion at webhook_converter.go:177
k8s.io/kubernetes/vendor/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion.(*safeConverterWrapper).ConvertToVersion at converter.go:151
k8s.io/kubernetes/vendor/k8s.io/apiextensions-apiserver/pkg/apiserver.schemaCoercingConverter.ConvertToVersion at customresource_handler.go:794
<autogenerated>:2
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer/versioning.(*codec).Encode at versioning.go:219
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime.Encode at codec.go:46
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime.CheckCodec at codec_check.go:31
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/cacher.NewCacherFromConfig at cacher.go:202
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/generic/registry.StorageWithCacher.func1 at storage_factory.go:65
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/generic/registry.(*Store).CompleteWithOptions at store.go:1385
k8s.io/kubernetes/vendor/k8s.io/apiextensions-apiserver/pkg/registry/customresource.newREST at etcd.go:102
k8s.io/kubernetes/vendor/k8s.io/apiextensions-apiserver/pkg/registry/customresource.NewStorage at etcd.go:44
k8s.io/kubernetes/vendor/k8s.io/apiextensions-apiserver/pkg/apiserver.(*crdHandler).getOrCreateServingInfoFor at customresource_handler.go:491
k8s.io/kubernetes/vendor/k8s.io/apiextensions-apiserver/pkg/apiserver.(*crdHandler).ServeHTTP at customresource_handler.go:217
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux.(*pathHandler).ServeHTTP at pathrecorder.go:248
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux.(*PathRecorderMux).ServeHTTP at pathrecorder.go:234
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server.director.ServeHTTP at handler.go:154
<autogenerated>:2
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithAuthorization.func1 at authorization.go:64
net/http.HandlerFunc.ServeHTTP at server.go:1947
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.WithMaxInFlightLimit.func1 at maxinflight.go:160
net/http.HandlerFunc.ServeHTTP at server.go:1947
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithImpersonation.func1 at impersonation.go:50
net/http.HandlerFunc.ServeHTTP at server.go:1947
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithAuthentication.func1 at authentication.go:81
net/http.HandlerFunc.ServeHTTP at server.go:1947
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP.func1 at timeout.go:100
runtime.goexit at asm_amd64.s:2361
```

it is for a test that is not exists in this PR but the line numbers should match enough.

sttts replied:
It's the codec smoke test. Should we have a way to disable that or alternatively to provide a smoke test object instead of an empty one?

mbohlool replied:
Depends on how the other discussion goes, this may not be important for CRD webhook conversion. I and @liggitt think a conversion from an empty object in one version should result in an empty object in another version. Let's conclude that discussion before going into this path.

mbohlool commented:
@deads2k @liggitt @sttts @lavalamp @roycaihw This is ready for review. PTAL.

sttts commented on test/images/crd-conversion-webhook/config.go:
```diff
@@ -0,0 +1,51 @@
+	return &tls.Config{
+		Certificates: []tls.Certificate{sCert},
+		// TODO: uses mutual tls after we agree on what cert the apiserver should use.
+		// ClientAuth:   tls.RequireAndVerifyClientCert,
```
This is a blocker for merging, right? 

What is it for admission by default? `--proxy-client-{cert,key}-file`? I remember we had a way to define a kubeconfig used instead.

mbohlool replied:
I am not sure about this, but why is it blocker when it is working? I copied this from webhook code and e2e test works at least locally.

sttts replied:
Right, it's only in the test image. Do we have certificate checks enable or disabled now in e2e tests?

mbohlool replied:
I did not enable certificate test. If is the same state as admission webhooks. How can we check if it is enabled or not?

sttts commented on test/images/crd-conversion-webhook/converter/example_converter.go:
```diff
@@ -0,0 +1,70 @@
+			host, hasHost := convertedObject.Object["host"]
+			port, hasPort := convertedObject.Object["port"]
+			if hasHost || hasPort {
+				convertedObject.Object["hostPort"] = fmt.Sprintf("%s:%s", host, port)
```
what does %s do with nil?

sttts commented on test/images/crd-conversion-webhook/converter/framework.go:
```diff
@@ -0,0 +1,138 @@
+
+// toConversionResponse is a helper function to create an AdmissionResponse
+// with an embedded error
+func toConversionResponse(msg string) *v1beta1.ConversionResponse {
```
toConversionFailure

mbohlool replied:
made it conversionResponseFailureWithMessagef (with f at the end for formatted version)

sttts commented on test/images/crd-conversion-webhook/converter/framework.go:
```diff
@@ -0,0 +1,138 @@
+		cr := unstructured.Unstructured{}
+		if err := cr.UnmarshalJSON(obj.Raw); err != nil {
+			glog.Error(err)
+			return toConversionResponse(fmt.Sprintf("failed to unmarshall object (%v) with error %v", string(obj.Raw), err))
```
with error:

sttts commented on test/images/crd-conversion-webhook/converter/framework.go:
```diff
@@ -0,0 +1,138 @@
+		}
+		var convertedCR *unstructured.Unstructured
+		var status metav1.Status
+		convertedCR, status = convert(&cr, convertRequest.DesiredAPIVersion)
```
`convertedCR, status := `

sttts commented on test/images/crd-conversion-webhook/converter/framework.go:
```diff
@@ -0,0 +1,138 @@
+
+func serve(w http.ResponseWriter, r *http.Request, convert convertFunc) {
+	var body []byte
+	if r.Body != nil {
```
never seen that check. Can that ever happen? Does ReadAll care?

sttts commented on test/images/crd-conversion-webhook/converter/framework.go:
```diff
@@ -0,0 +1,138 @@
+	// verify the content type is accurate
+	contentType := r.Header.Get("Content-Type")
+	if contentType != "application/json" {
+		err := fmt.Errorf("contentType=%s, expect application/json", contentType)
```
`msg := fmt.Sprintf`

sttts commented on test/images/crd-conversion-webhook/converter/framework.go:
```diff
@@ -0,0 +1,138 @@
+		return
+	}
+
+	glog.V(2).Info(fmt.Sprintf("handling request: %v", body))
```
Infof

sttts commented on test/e2e/apimachinery/crd_conversion_webhook.go:
```diff
@@ -0,0 +1,297 @@
+)
+
+// TODO(mehdy): this should be 1.13.0
+var serverCRDConversionWebhookVersion = utilversion.MustParseSemantic("v1.12.0")
```
how and when is this updated? After the alpha tagging?

mbohlool replied:
actually this should work now. Will update it to 1.13.0

sttts commented on test/e2e/apimachinery/crd_conversion_webhook.go:
```diff
@@ -0,0 +1,297 @@
+		cleanCRDWebhookTest(client, namespaceName)
+	})
+
+	It("Should be able to convert from CRD v1 to CRD v2", func() {
```
CR v1 to CR v2

sttts commented on cmd/kube-apiserver/app/apiextensions.go:
```diff
@@ -74,12 +77,14 @@ func createAPIExtensionsConfig(
 }
 
-func createAPIExtensionsServer(apiextensionsConfig *apiextensionsapiserver.Config, delegateAPIServer genericapiserver.DelegationTarget) (*apiextensionsapiserver.CustomResourceDefinitions, error) {
+func createAPIExtensionsServer(apiextensionsConfig *apiextensionsapiserver.Config, serviceResolver webhook.ServiceResolver, authResolverWrapper webhook.AuthenticationInfoResolverWrapper, delegateAPIServer genericapiserver.DelegationTarget) (*apiextensionsapiserver.CustomResourceDefinitions, error) {
```
the two args are not used anymore.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go:
```diff
@@ -78,6 +79,11 @@ type ExtraConfig struct {
 	// the CRD Establishing will be hold by 5 seconds.
 	MasterCount int
+
+	// ServiceResolver is used in CR webhook converters to resolve webhook's Service names
```
nit: _s_ervice names

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go:
```diff
@@ -119,7 +125,7 @@ func (cfg *Config) Complete() CompletedConfig {
 
 // New returns a new instance of CustomResourceDefinitions from the given config.
-func (c completedConfig) New(delegationTarget genericapiserver.DelegationTarget) (*CustomResourceDefinitions, error) {
+func (c CompletedConfig) New(delegationTarget genericapiserver.DelegationTarget) (*CustomResourceDefinitions, error) {
```
don't make this public. It's private intentionally.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go:
```diff
@@ -0,0 +1,272 @@
+	}, nil
+}
+
+// isEmptyUnstructuredObject returns true of in is an empty unstructured object, i.e. an unstructured object that does
```
s/of/if/

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go:
```diff
@@ -0,0 +1,272 @@
+
+// isEmptyUnstructuredObject returns true of in is an empty unstructured object, i.e. an unstructured object that does
+// not have any field except apiVersion and kind.
+func isEmptyUnstructuredObject(in runtime.Object) bool {
```
nit: move helpers to the file end

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go:
```diff
@@ -0,0 +1,272 @@
+// isEmptyUnstructuredObject returns true of in is an empty unstructured object, i.e. an unstructured object that does
+// not have any field except apiVersion and kind.
+func isEmptyUnstructuredObject(in runtime.Object) bool {
+	Object, ok := in.(*unstructured.Unstructured)
```
s/Object/u/

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go:
```diff
@@ -0,0 +1,272 @@
+}
+
+func (c *webhookConverter) ConvertToVersion(in runtime.Object, target runtime.GroupVersioner) (runtime.Object, error) {
+	if isEmptyUnstructuredObject(in) {
```
when does this happen? Why should we handle that in a special way?

mbohlool replied:
actually now that we fixed the example webhook and also expect conversion webhooks to not fail on empty objects, this is less important. But it happens. This is where it initiated:

https://github.com/kubernetes/kubernetes/blob/dbb448bbdcb9e440eee57024ffa5f1698956a054/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L201

The config.Type comes from a `NewFunc` which creates an empty object and this line will pass that eventually to webhook converter to convert an empty object from one version to another.

I don't think we should call webhooks on an empty object as the behaviour should be like NopConverter. Generally conversion should not fail based on fields not exists. That is validation problem not conversion. Also we do not expect conversion webhook to do any defaulting too. So there is no point calling webhook on empty object.

Because we consistently create these empty objects in our machinery logic, it make sense to special case it here.

sttts replied:
It's some smoke test we added to the storage plumbing. @liggitt @deads2k do you remember why?

In theory an empty object in one version could map to a non-empty object in another version. So strictly speaking this special casing is wrong.

mbohlool replied:
'In theory an empty object in one version could map to a non-empty object in another version. So strictly speaking this special casing is wrong.'

how? do you have an example?

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go:
```diff
@@ -0,0 +1,272 @@
+
+	request := createConversionReview(in, toGVK.GroupVersion().String())
+	response := &v1beta1.ConversionReview{}
+	ctx := context.TODO()
```
should we put a timeout here? What does admission do?

mbohlool replied:
It is the same there:
https://github.com/kubernetes/kubernetes/blob/5652d5cffadcd8a2f107b6aecf5fc06c0fc473f1/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/webhook.go#L167

sttts replied:
Remembering the discussion to put one second here or something like that. But if admission is not better, this is not critical. But please add a TODO (also to admission while you are on it).

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/server.go:
```diff
@@ -55,10 +55,6 @@ func Run(o *options.CustomResourceDefinitionsServerOptions, stopCh <-chan struct
 	}
-
 	server, err := config.Complete().New(genericapiserver.NewEmptyDelegate())
-	if err != nil {
```
why?

sttts commented on staging/src/k8s.io/apiserver/pkg/util/webhook/auth_resolver_wrapper.go:
```diff
@@ -0,0 +1,50 @@
+)
+
+// CreateWebhookAuthResolverWrapper creates a auth wrapper using the loopback and proxyTransport configs.
+func CreateWebhookAuthResolverWrapper(LoopbackClientConfig *restclient.Config, proxyTransport *http.Transport) AuthenticationInfoResolverWrapper {
```
didn't we have that already for the admission case? Can we reuse it?

mbohlool commented:
/test pull-kubernetes-e2e-kops-aws

mbohlool commented on test/images/crd-conversion-webhook/converter/framework.go:
```diff
@@ -0,0 +1,136 @@
+	// reset the request, it is not needed in a response.
+	convertReview.Request = &v1beta1.ConversionRequest{}
+
+	resp, err := json.Marshal(convertReview)
```
assumed the output content type to be the same as input. Let me know if this is a bad assumption.

mbohlool commented:
@sttts @liggitt Please take another look.

mbohlool commented:
/retest

mbohlool commented:
/test pull-kubernetes-cross
/test pull-kubernetes-integration

sttts commented:
Lgtm. Waiting for @liggitt to give the final lgtm.

mbohlool commented:
/test pull-kubernetes-e2e-kops-aws
/test pull-kubernetes-e2e-gce-100-performance

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go:
```diff
@@ -0,0 +1,278 @@
+			if err != nil {
+				return nil, err
+			}
+			if err := ensureCorrectGVK(converted, toGVK); err != nil {
```
nit: include the index or some identifying info about the invalid object if available (uid, name, namespace, etc)? just trying to think what info I would want to know if I got this error

liggitt replied:
same goes for all the errors encountered inside this loop

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go:
```diff
@@ -0,0 +1,278 @@
+		return nil, fmt.Errorf("conversion request failed for %v, Response: %v", in.GetObjectKind(), response)
+	}
+
+	if len(response.Response.ConvertedObjects) != len(request.Request.Objects) {
```
how paranoid do we want to be about making sure the converted objects match their inputs (kind, uid, name, namespace must not change between the input and output)? that would make me more confident that a buggy or compromised conversion webhook wasn't causing data to be written to the wrong place in etcd, etc

sttts replied:
For lists we are paranoid. Why not for non-lists as well, with the same `validateConvertedObject` validation?

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go:
```diff
@@ -0,0 +1,278 @@
+		return rx.Object, nil
+	}
+	u := unstructured.Unstructured{}
+	err := u.UnmarshalJSON(rx.Raw)
```
surprised `u` doesn't have to be a pointer here

mbohlool replied:
code:

```golang
// You can edit this code!
// Click here and start typing.
package main

import "fmt"

type A struct {value int}
type AP struct {value int}

func (a A) set(v int) { a.value = v}
func (a *AP) set(v int) { a.value = v}
func (a A) get() int { return a.value }
func (a *AP) get() int { return a.value }

func main() {
        a := A{}
        ap := AP{}
        b := &A{}
        bp := &AP{}
        a.set(10)
        ap.set(12)
        b.set(14)
        bp.set(16)
	fmt.Println(a.get())
	fmt.Println(ap.get())
	fmt.Println(b.get())
	fmt.Println(bp.get())
}
```

output:

```
0
12
0
16
```

which is interesting but in our case, we are fine.

sttts replied:
Why should it be a pointer? The value `u` is addessible (is that the right term in Go semantics) such that the referencing is done implicitly by Go.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -147,10 +152,15 @@ func NewCustomResourceDefinitionHandler(
 			ret.removeDeadStorage()
 		},
 	})
+	CRConverterFactory, err := conversion.NewCRConverterFactory(serviceResolver, authResolverWrapper)
```
nit: `crConverterFactory`

mbohlool commented:
@liggitt addressed your comments. PTAL.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go:
```diff
@@ -0,0 +1,303 @@
+	return &webhookConverterFactory{clientManager}, nil
+}
+
+// webhookConverter is a converter calls an external webhook to do the CR conversion.
```
that calls

sttts commented on test/images/crd-conversion-webhook/converter/framework.go:
```diff
@@ -0,0 +1,178 @@
+
+	accept := r.Header.Get("Accept")
+	outSerializer := getOutputSerializer(accept)
+	if outSerializer == nil {
```
outSerializer is not used. Intentional?

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go:
```diff
@@ -0,0 +1,303 @@
+		return fmt.Errorf("request to convert CR to an invalid group/version: %s", inGVK.String())
+	}
+
+	converted, err := c.ConvertToVersion(unstructIn.DeepCopy(), outGVK.GroupVersion())
```
why the deepcopy?

mbohlool replied:
generally ConvertToVersion can change input but in our case we don't need it.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go:
```diff
@@ -0,0 +1,303 @@
+	unstructuredConverted, ok := converted.(runtime.Unstructured)
+	if !ok {
+		// this should not happened
+		return fmt.Errorf("internal server error, CR conversion failed")
```
This "internal server error" is added implicitly by the caller, isn't it? Looks strange here.

mbohlool replied:
dropped all "internal server error" in this file.

mbohlool commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go:
```diff
@@ -0,0 +1,303 @@
+		return c.nopConverter.ConvertToVersion(in, target)
+	}
+
+	fromGVK := in.GetObjectKind().GroupVersionKind()
```
very nice catch.

I already have an integration test that confirms this works fine for nopConverter.

for webhookConverter, it was a bug. the check for equal GVKs should not happened like this. I already fixed it and added an e2e test for it to make sure it works. Will send out the update when it passed.

liggitt replied:
`fromGVK := in.GetObjectKind().GroupVersionKind()` still seems problematic if `in` is a v1 List. Is that expected to occur?

mbohlool replied:
We only care about its GV not kind. I've cleaned up the code to make sure only `toGV` will move forward from this part of the code.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go:
```diff
@@ -0,0 +1,341 @@
+	if isList {
+		for i, item := range listObj.Items {
+			if !c.validVersions[item.GroupVersionKind().GroupVersion()] {
+				return nil, fmt.Errorf("request to convert List of CRs to an invalid group/version for item %v: %s", i, item.GroupVersionKind())
```
this message is confusing... `item.GroupVersionKind()` is the version we're converting *from*, not to

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go:
```diff
@@ -0,0 +1,341 @@
+				return nil, fmt.Errorf("request to convert List of CRs to an invalid group/version for item %v: %s", i, item.GroupVersionKind())
+			}
+		}
+		if !c.validVersions[fromGVK.GroupVersion()] {
```
either compute fromGVK per-item inside the loop or move this check and message outside the loop. isn't `fromGVK` for a particular item `item.GroupVersionKind()`?

mbohlool replied:
This check was not necessary at all.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go:
```diff
@@ -0,0 +1,341 @@
+				continue
+			}
+			converted, err := getRawExtensionObject(response.Response.ConvertedObjects[convertedIndex])
+			original := request.Request.Objects[convertedIndex].Object
```
use `listObj.Items[i]` as the original for comparison since that's what we're replacing with the converted copy on line 275

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go:
```diff
@@ -0,0 +1,341 @@
+			}
+			convertedList.Items[i] = *unstructConverted
+		}
+		convertedList.SetAPIVersion(toGVK.GroupVersion().String())
```
if the input was a `v1` List containing `example.com/v1alpha1` Widget items, requesting conversion to `example.com/v1beta1`, wouldn't this result in an `example.com/v1alpha1` List as output?

mbohlool replied:
well, it depends what `target.KindForGroupVersionKinds` meant to do. For GroupVersion.target.KindForGroupVersionKinds (which I think the type of target we pass to conversion is GroupVersion right now) it will only replace kind with the kind of given GVK, so 

```
toGVK, ok := target.KindForGroupVersionKinds([]schema.GroupVersionKind{fromGVK})
```

target is (example.com, v1beta1), fromGVK is (example.com, v1, ListType), then the toGVK will be (example.com, v1beta1, List) and as we only use GroupVersion here, it will be the right APIVersion example.com/v1beta1.

How can any of these being affected by the list item's APIVersion?

mbohlool replied:
I cleared the code up to only have toGV moving forward from initial checks. We don't need anything else for conversion. `toGV.String()` is `APIVersion`.

liggitt replied:
That was why I asked what types of lists we support passing to this conversion function. We have two types of lists that get passed around:
* generic `v1` List objects containing non-homogenous objects of any arbitrary GVK. If you were going to read a bunch of objects directly from storage into a list, it would be this type (because storage does not guarantee all stored objects are in the same GVK). For this type of list, using the list's GVK to compute to/from for the items is incorrect, and setting the requested target version as the list's GV is also incorrect.
* specific GVK list objects (like `example.com/v1beta1` `MyKindList`) containing only objects with the same GVK (`example.com/v1beta1` `MyKind`). For this type of list, using the GV of the List to compute to/from for the items is more correct, and setting the requested target version as the list's GV would also be correct.

The analogue for built-in types would be:

Generic list:
* in:
    ```
    kind: List
    apiVersion: v1
    items:
    - {"kind":"Deployment","apiVersion":"apps/v1beta1",...}
    - {"kind":"Deployment","apiVersion":"apps/v1beta2",...}
   ```
* converted to `extensions/v1beta1`
* out:
    ```
    kind: List
    apiVersion: v1
    items:
    - {"kind":"Deployment","apiVersion":"extensions/v1beta1",...}
    - {"kind":"Deployment","apiVersion":"extensions/v1beta1",...}
    ```

Specific list:
* in:
    ```
    kind: DeploymentList
    apiVersion: apps/v1beta1
    items:
    - {"kind":"Deployment","apiVersion":"apps/v1beta1",...}
    - {"kind":"Deployment","apiVersion":"apps/v1beta1",...}
    ```
* converted to `extensions/v1beta1`
* out:
    ```
    kind: DeploymentList
    apiVersion: extensions/v1beta1
    items:
    - {"kind":"Deployment","apiVersion":"extensions/v1beta1",...}
    - {"kind":"Deployment","apiVersion":"extensions/v1beta1",...}
    ```

liggitt replied:
Talked offline. v1 List objects aren't passed here (and if they were, the validVersions[fromGVK] check would fail and prevent insanity).

CRD storage impl creates an unstructured list with the request's GV, populates it from storage, then calls conversion to convert the individual items.

Specific list:
* in:
    ```
    kind: MyKindList
    apiVersion: acme.com/v1
    items:
    - {"kind":"MyKind","apiVersion":"acme.com/v1",...}
    - {"kind":"MyKind","apiVersion":"acme.com/v1beta1",...}
    ```
* converted to `acme.com/v1`
* out:
    ```
    kind: MyKindList
    apiVersion: acme.com/v1
    items:
    - {"kind":"MyKind","apiVersion":"acme.com/v1",...}
    - {"kind":"MyKind","apiVersion":"acme.com/v1",...}
    ```

this differs from built-in resources which cannot contain non-homogenous kinds in a specifically typed list

mbohlool replied:
CRDs do not use generic list. The list always has the type of the request and it is different than standard type codepath. Decoder will put objects with different GVK in an unstructuredList. The unstructuredList is created by the request store object which will have the GVK of the request. Thus here we assume even if the input is the list, the GV is what we want.
We are protected against the generic list because we check for GVK to be valid at the start of the conversion and a generic v1 list will not pass that.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go:
```diff
@@ -0,0 +1,341 @@
+	var objects []runtime.RawExtension
+	if isList {
+		for _, item := range listObj.Items {
+			// Only sent item for conversion, if the apiVersion is different
```
this has the possibility of producing a zero-length objects array. skip conversion call in that case?

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go:
```diff
@@ -0,0 +1,341 @@
+		for _, item := range listObj.Items {
+			// Only sent item for conversion, if the apiVersion is different
+			if item.GetAPIVersion() != apiVersion {
+				objects = append(objects, runtime.RawExtension{Object: item.DeepCopyObject()})
```
why is deep copy required?

mbohlool committed:
CRD Conversion

mbohlool commented:
/test pull-kubernetes-e2e-gce-100-performance

liggitt commented:
/lgtm
/approve

liggitt commented:
`staging/src/k8s.io/apiextensions-apiserver` needs approval by @sttts

sttts commented:
/approve
